File: 0382-linked-list-random-node/0382-linked-list-random-node.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    private ListNode head;
    private Random rand;

    public Solution(ListNode head) {
        this.head = head;
        this.rand = new Random();
    }
    
    public int getRandom() {
        ListNode curr = head;
        int result = -1;
        int count = 0;
        
        while (curr != null) {
            count++;
            // Probability of selecting current node: 1/count
            if (rand.nextInt(count) == 0) {
                result = curr.val;
            }
            curr = curr.next;
        }
        
        return result;
    }
}

File: 0382-linked-list-random-node/0382-linked-list-random-node.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    ListNode* head;
public:
    /** @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node. */
    Solution(ListNode* head) : head(head) {
        srand(time(NULL)); // Optional for better randomness
    }
    
    /** Returns a random node's value. */
    int getRandom() {
        int result = -1;
        int count = 0;
        ListNode* curr = head;
        
        while (curr) {
            count++;
            // Replace result with current node with probability 1/count
            if (rand() % count == 0) {
                result = curr->val;
            }
            curr = curr->next;
        }
        
        return result;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(head);
 * int param_1 = obj->getRandom();
 */